internal/worker/sshserver/k8s/session_test.go

```diff
@@ -79,3 +85,88 @@ func (s *k8sSessionSuite) TestSessionHandler(c *gc.C) {
 	k8sHandlers.SessionHandler(s.session)
 	c.Assert(readWriter.String(), gc.Equals, "failed to resolve k8s exec info: error\n")
 }
+
+func (s *k8sSessionSuite) TestSessionHandlerPty(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	l := loggo.GetLogger("test")
+	s.resolver.EXPECT().ResolveK8sExecInfo(gomock.Any()).Return(params.SSHK8sExecResult{
+		PodName:   "test-pod",
+		Namespace: "test-namespace",
+	}, nil)
+	s.executor.EXPECT().Exec(gomock.Any(), gomock.Any()).DoAndReturn(func(params k8sexec.ExecParams, opts <-chan struct{}) error {
+		c.Assert(params.PodName, gc.Equals, "test-pod")
+		c.Assert(params.ContainerName, gc.Equals, "test-container")
+
+		// Simulate a write
+		_, err := params.Stdout.Write([]byte("test output"))
+		c.Assert(err, jc.ErrorIsNil)
+		return nil
+	})
+
+	virtualHostame, err := virtualhostname.NewInfoContainerTarget(uuid.New().String(), "test/0", "test-container")
+	c.Assert(err, jc.ErrorIsNil)
+
+	k8sHandlers, err := NewHandlers(
+		virtualHostame,
+		s.resolver,
+		l,
+		func(string) (k8sexec.Executor, error) {
+			return s.executor, nil
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	s.session.EXPECT().Environ()
+	s.session.EXPECT().Command()
+	s.context.EXPECT().Done()
+	s.session.EXPECT().Context().Return(s.context)
+	mockSession := userSession{
+		Session: s.session,
+		isPty:   true,
+	}
+	// Simulate the user closing the session.
+	go func() {
+		time.Sleep(testing.ShortWait)
+		mockSession.Close()
+	}()
+	k8sHandlers.SessionHandler(&mockSession)
+	c.Assert(err, jc.ErrorIsNil)
+	// we can't use Equals because the pty device generates a lot of characters.
+	c.Assert(mockSession.stdout.String(), jc.Contains, "test output")
+	c.Assert(mockSession.stderr.String(), gc.Equals, "")
+	c.Assert(mockSession.stdin.String(), gc.Equals, "")
+}
+
+type userSession struct {
+	ssh.Session
+	stdin  bytes.Buffer
+	stdout bytes.Buffer
+	stderr bytes.Buffer
+	isPty  bool
+	closed bool
+}
+
+func (u *userSession) Write(p []byte) (n int, err error) {
+	return u.stdout.Write(p)
+}
+
+// Read is not returning EOF to similate an interactive session.
+func (u *userSession) Read(p []byte) (n int, err error) {
+	if u.closed {
+		return 0, io.EOF
+	}
+	p = []byte("command")
+	return len(p), nil
```

Comment:
This looks wrong at first glance. If something calls `Read()` in a loop, this will cause high CPU usage. This should block on a channel or something similar until you signal it to return.